File: dotnet/Lexy.Poc/Lexy.Poc.Core/Language/Expressions/Expression.cs
using System;
using System.Collections.Generic;
using System.IO;
using Lexy.Poc.Core.Parser;

namespace Lexy.Poc.Core.Language.Expressions
{
    public abstract class Expression : Node
    {
        public ExpressionSource Source { get; }

        protected Expression(ExpressionSource source)
        {
            Source = source ?? throw new ArgumentNullException(nameof(source));
        }

        public override string ToString()
        {
            var writer = new StringWriter();
            foreach (var token in Source.Tokens)
            {
                writer.Write(token.Value);
            }
            return writer.ToString();
        }
    }

    public class ExpressionSource
    {
        public Line Line { get; }
        public TokenList Tokens { get; }

        public ExpressionSource(Line line, TokenList tokens)
        {
            Line = line;
            Tokens = tokens;
        }
    }
}

File: typescipt/src/language/expressions/expression.ts


export class Expression extends Node {
   public ExpressionSource Source

   protected Expression(ExpressionSource source, SourceReference reference) : base(reference) {
     Source = source ?? throw new Error(nameof(source));
   }

   public override toString(): string {
     let writer = new StringWriter();
     foreach (let token in Source.Tokens) writer.Write(token.Value);
     return writer.ToString();
   }

   public abstract deriveType(context: IValidationContext): VariableType;
}
